### PR TITLE
Validate the enum type matches the `-type` arg

### LINKED ISSUE #24: Mixed enum values when for two distinct enum types in the same file
If there are multiple enum types in a single file, the generated code combines all their (string)  names.

It seems that problem is that the `-type` argument is ignored; It's probably related to the comment in the code:
```go
		// TODO: check that type matches
```

Try this file for example:

```go
package main

type Enum1 uint8

//go:generate go-enum-encoding -type=Enum1
const (
    OnePlain Enum1 = iota+1 // json:"one_plain"
)

type Enum2 uint8

//go:generate go-enum-encoding -type=Enum2
const (
    TwoPlain Enum2 = iota+1 // json:"two_plain"
)

func main() {
    println("hello world")
}
```

### PR DESCRIPTION
**Pull request by orenl**
Without validation, multiple definitions of multiple enums in the same file would be mixed, as well as comments with `json:"..."` on `val` or `const` statments.

Fixes #24

### CODE REVIEW

**Comment by nikolaydubina** [comment]
nice. let me read through it

File: main.go [hunk 1]
```diff
@@ -57,27 +57,43 @@ func process(typeName string, fileName string, packageName string, mode string)
 
 	var specs [][2]string
 
-	ast.Inspect(f, func(node ast.Node) bool {
-		spec, ok := node.(*ast.ValueSpec)
-		if !ok {
+	ast.Inspect(f, func(astNode ast.Node) bool {
+		node, ok := astNode.(*ast.GenDecl)
+		if !ok || node.Tok != token.CONST {
```
Comment (nikolaydubina) [inline main.go]: will this work with?

```go
var (

....
)
```

File: main.go [hunk 2]
```diff
@@ -57,27 +57,43 @@ func process(typeName string, fileName string, packageName string, mode string)
...
-		}
-		// TODO: check that type matches
+		var typeFound *ast.Ident
 
-		tag, ok := "", false
-		for _, field := range strings.Fields(spec.Comment.Text()) {
-			if strings.HasPrefix(field, "json:") {
-				tag, ok = field[len("json:\""):len(field)-1], true
-				break
+		for _, astSpec := range node.Specs {
```
Comment (nikolaydubina) [inline main.go:77]: yeah, something like this. but I would suspected you need tree walking here.

if it is available in direct children, thats perfect.

File: main.go [hunk 3]
```diff
@@ -57,27 +57,43 @@ func process(typeName string, fileName string, packageName string, mode string)
...
+				continue
+			}
+
+			if len(spec.Names) != 1 {
+				return false
+			}
+
+			if typeFound == nil {
+				typeFound, _ = spec.Type.(*ast.Ident)
+				if typeFound == nil || typeFound.Name != typeName {
```
Comment (nikolaydubina) [inline main.go]:
```suggestion
				if typeFound, _ = spec.Type.(*ast.Ident); typeFound == nil || typeFound.Name != typeName {
```

**Review by nikolaydubina — COMMENTED** [review]
looks good. please check failing test

+also add test case that you mentioned in the issue that would be failing without this PR.

**Comment by orenl** [comment]
> looks good. please check failing test

The test fails because the _first_ enum constant doesn't explicitly state its type (which is fair play of course):
```
    UndefinedColor = Color{}            // json:""
```
Rather than:
```
    UndefinedColor Color = Color{}            // json:""
```
 
So in the case that the isn't an explicit type for the _first_ enum constant, then:
~~If the type is `struct`, or an literal, it's easy to infer; I will fix that.~~
If the type is `struct`, it's easy to infer, and the result will usually will be the developer's intent (unlikely to have two sets of enums constants with the same dedicated struct type).
If the type is a plain literal, it's possible to infer (a bit more code), but the result may be not what the developer intended: consider in the same file two types both alias to `uint8`, and neither states the enum type explicitly, so impossible to tell.
~~However,~~ if the type is an expression (like `2 * iota`) then it gets involved - unsure how to do easily. So in that case I'd ignore and skip (rather than silently continue and hope there is only one enum in the file).

**Comment by orenl** [comment]
Let me rephrase the analysis - consider this code:

```go
type Enum1 uint8

//go:generate go-enum-encoding -type=Enum1
const (
    Plain_1a Enum1 = iota+1 // json:"plain_1a"        <-- first enum states explicit type
    Plain_1b                // json:"plain_1b"
)

type Enum2 struct { c uint8 }

//go:generate go-enum-encoding -type=Enum2
const (
    Plain_2a = Enum2{1} // json:"plain_2a"            <-- first enum uses struct type (no explicit type)
    Plain_2b            // json:"plain_2b"
)

type Enum3 uint8

//go:generate go-enum-encoding -type=Enum3
const (
    Plain_3a = 1 // json:"plain_3a"                   <-- first enum uses literal (no explicit type)
    Plain_3b = 2 // json:"plain_3b"
)

type Enum4 uint8

//go:generate go-enum-encoding -type=Enum4
const (
    Plain_4a = iota + 1 // json:"plain_4a"            <-- first enum uses expression (no explicit type)
    Plain_4b = 2        // json:"plain_4b"
)
```

1. For `Enum1` the explicit enum type can be extracted directly
2. For `Enum2` the implicit enum type can be inferred from the composite literal (struct)
3. For `Enum3` the implicit enum cannot be inferred (literal type is available with extra work)
4. For `Enum4` the implicit enum cannot be inferred (and challenging to infer the literal type)

For case 1,2, the problem is easily solved - I already have it fixed here.
For case 3 and 4, it is not possible distinguish between two enum sets in the same file.

For cases 3 and 4 there I can think of two options:
1. Skip and do not generate code for such enum constants, and emit useful error for the developer.
2. Ignore the problem - the generated code will fail to compile at best, or work wrongly at worst.

My choice is no. 1, but I thought I'd get feedback before I go at it.

**Comment by orenl** [comment]
With the new code in place, I now see a simpler way to do this: always consider the const/var block immediately subsequent to the `#go:generate ...` line, whose location we know. That rids the need to detect or infer the type of enum values.

File: main.go (hunk 1 above)
Comment (orenl) [reply main.go]: Good point. I wrong assumed `var` would not be used for enums.
Fixed.

File: main.go (hunk 2 above)
Comment (orenl) [reply main.go:77]: At this point we already know we are in `const` or `var`, and seems like in the test cases it's available in direct children. Can you think of an example where tree walking would be needed?

File: main.go (hunk 3 above)
Comment (orenl) [reply main.go]: Obsolete (code removed)

File: main.go (hunk 2 above)
Comment (nikolaydubina) [reply main.go:77]: no, it was just a hunch

**Comment by nikolaydubina** [comment]
>   UndefinedColor = Color{}            // json:""

you are right. make sense

> const (
> Plain_4a = iota + 1 // json:"plain_4a"   
> and
> const (
> Plain_3a = 1 // json:"plain_3a" 

untyped consts, well, they dont have type. so either type should be present for us or we detect by `go:generate` location in file as you proposed

> using go:generate location + GOLINE

nice. very good actually. let me check through it. but very likely this is good to merge.

File: main.go [hunk 4]
```diff
@@ -50,37 +52,58 @@ func process(typeName string, fileName string, packageName string, mode string)
...
-		if ok {
-			specs = append(specs, [2]string{spec.Names[0].Name, tag})
-		}
+
 		return false
 	})
 
+	if len(specs) == 0 {
+		return errors.New(
+			fmt.Sprintf("%s: Unable to find values for enum type %q\n", fileName, typeName))
```
Comment (nikolaydubina) [inline main.go]:
```suggestion
		return fmt.Errorf("%s: unable to find values for enum type %q", fileName, typeName)
```

File: main.go [hunk 5]
```diff
@@ -50,37 +52,57 @@ func process(typeName string, fileName string, packageName string, mode string)
...
+	ast.Inspect(f, func(astNode ast.Node) bool {
+		node, ok := astNode.(*ast.GenDecl)
+		if !ok || (node.Tok != token.CONST && node.Tok != token.VAR) {
 			return true
 		}
 
-		if len(spec.Names) != 1 {
+		position := fset.Position(node.Pos())
+		if position.Line != expectedLine {
 			return false
```
Comment (nikolaydubina) [inline main.go:74]: guess you need true here. we did not find position in AST yet, so it may or may not be inside subtree. good to iterate further. this is akin to continuing tree walking until detection of var/const declaration above. (can't think example now yet though).
```suggestion
			return true
```

File: internal/testdata/size_enum_encoding.go [hunk 6]
```diff
@@ -1,39 +0,0 @@
-// Code generated by go-enum-encoding DO NOT EDIT
```
Comment (nikolaydubina) [inline internal/testdata/size_enum_encoding.go]: why this file was removed? can you keep it?

Comment (nikolaydubina) [reply internal/testdata/size_enum_encoding.go]: oh, I see you moving expected files/output of generation to exp folder. ok

File: main.go (hunk 5 above)
Comment (orenl) [reply main.go:74]: I believe false is correct here, since at this point we know already know we are in const/var block, so if location in the file does not match then there is not point in continuing with children. The traversal will retreat one level up and proceed from there.

File: internal/testdata/size_enum_encoding.go (hunk 6 above)
Comment (orenl) [reply internal/testdata/size_enum_encoding.go]: I consider the files in `internal/testdata/exp` the source of truth for the test, while the output files in `internal/testdata` are generated by the running tests and should be temporary. In fact, we should clean them up before/after the test - otherwise a test that failed to generate anything and also failed to report an error - will pass. What do you think?

Specifically for `internal/testdata/size.go`, it looks like it is/was not being used for testing to begin with. Maybe left out unintentionally?

File: main.go (hunk 5 above)
Comment (nikolaydubina) [reply main.go:74]: right.

File: internal/testdata/size_enum_encoding.go (hunk 6 above)
Comment (nikolaydubina) [reply internal/testdata/size_enum_encoding.go]: ok. all looks good. 

but please don't delete files. I hear you point, it makes sense to have exp files that actually used in tests. let's do that later. for now better to keep those files.

**Review by nikolaydubina — APPROVED** [review]
ok merging.

@orenl very good work!

image
